slice/lis.go

```diff
@@ -134,6 +135,72 @@ func LNDSFunc[T any, Slice ~[]T](vs Slice, cmp func(a, b T) int) Slice {
 	return ret
 }
 
+// LISFunc computes a longest strictly increasing subsequence of vs.
+//
+// This implementation takes O(P·log(n)) time and O(n) space for
+// inputs of length n = len(vs) and longest subsequence length P. If
+// the longest subsequence is the entire input, it takes O(n) time and
+// O(n) space.
+func LIS[T cmp.Ordered, Slice ~[]T](vs Slice) Slice {
+	return LISFunc(vs, cmp.Compare)
+}
+
+// LISFunc computes a longest strictly increasing subsequence of vs,
+// in the order determined by the cmp function. cmp must return a
+// negative number when a < b, a positive number when a > b, and zero
+// when a == b.
+//
+// This implementation takes O(P·log(n)) time and O(n) space for
+// inputs of length n = len(vs) and longest subsequence length P. If
+// the longest subsequence is the entire input, it takes O(n) time and
+// O(n) space.
+func LISFunc[T any, Slice ~[]T](vs Slice, cmp func(a, b T) int) Slice {
+	// LISFunc is almost exactly the same as LNDSFunc, except that
+	// comparisons are > instead of >= and the binary search leans
+	// left. Further comments have been omitted for brevity.
+	if len(vs) == 0 {
+		return vs
+	}
+
+	var (
+		tails = make([]int, 1, len(vs))
+		prev  = make([]int, len(vs))
+	)
+	prev[0] = -1
+	tails[0] = 0
+
+	for i := range vs[1:] {
+		i++
+
+		idxOfBestTail := tails[len(tails)-1]
+		if cmp(vs[i], vs[idxOfBestTail]) > 0 {
+			prev[i] = idxOfBestTail
+			tails = append(tails, i)
+			continue
+		}
+
+		replaceIdx, _ := slices.BinarySearchFunc(tails[:len(tails)-1], vs[i], func(idx int, target T) int {
```

Comment:
Maybe note that we are searching for an item known to be present, so ignoring the flag is safe?

Reply:
I don't think we do know that ti's present, but the same comment as in LNDS does apply: we're overwriting whatever slot we get back from this search, and because we're searching tails minus its final element, the returned index is always valid and the one we want to use.